brainstorming.md

```diff
@@ -0,0 +1,372 @@
+- all columns are of equal length
+
+All SOMA data objects are named with URIs.
+
+> ⚠️ **Issue** -- this section would benefit from a data model diagram, helpfing visualize the relationships between the types.
+
+> ⚠️ **Issue** -- the use of `SOMA` to name the API _and_ a data type is confusing. We should consider a new name for the aforementioned SOMA _type_. _Candidate:_ `SOMAAnnotatedMatrix`?
+
+## Base Type System
+
+The SOMA API borrows its base type system from the Arrow language-agnostic in-memory system for data typing and serialization ([format](https://arrow.apache.org/docs/format/Columnar.html)). The SOMA API is intended to be used with an Arrow implementation such as [PyArrow](https://arrow.apache.org/docs/python/) or the [Arrow R package](https://arrow.apache.org/docs/r/), or other libraries which interoperate with Arrow (eg, Pandas).
+
+Where SOMA requires an explicit typing system, it utilizes the Arrow types and schema. SOMA has no specific requirements for the type or serialization system used by the underlying storage engine, other than it be capable of understanding and representing the Arrow types. It is understood and expected that each implementation of SOMA will have data type expressiveness limits (eg, just because you can express a type in he Arrow type system does not mean all SOMA implementations will understand it).
+
+In the following doc:
+
+- `primitive` types in this specification refers to Arrow primitive types, eg, `int32`, `float`, etc.
+- `string` refers to Arrow UTF-8 variable-length `string`, ie, `List<Char>`.
+- `simple` types include all primitive types, plus `string`.
+
+Other Arrow types are explicitly noted as such, eg, `Arrow.RecordBatch`.
+
+> ⚠️ **Issue** - is the use of Arrow as a base type language acceptable? There are others, up to and including inventing a new system, but it seemed prudent to adopt an existing, language neutral specification, which is already supported by many "engines".
+
+> ⚠️ **Issue** - are there parts of the Arrow type system that we wish to _explicitly exclude_ from SOMA? I have left this issue open (ie, no specific text) for now, thinking that we can come back and subset as we understand what complex types are required, and how much flexibility should be in this spec. We clearly need some complex types (eg, RecordBatch, List, etc) as they are implied by `string`, etc. My own preference would be to mandate a small set of primitive types, and leave the rest open to implementations to support as they feel useful.
+
+## Metadata (uns)
+
+All SOMA objects may be annotated with a small amount of metadata, in the form of a simple key/value map:
+
+- `uns` (map[string, simple]) - string-keyed mapping to any **simple** type (eg, int32, string, etc).
+
+Only `simple` types are supported in the metadata values. The metadata lifecycle is the same as its containing object, eg, it will be deleted when the containing object is deleted.
```

Comment:
we need to say what 'simple' is. allowing a metadata value to be a list of strings seems really crucial -- is a list of strings 'simple'?

Reply:
`simple` is defined in the section titled `Base Type System`.  I will add a link in the spec to clarify.